state/validation.go

```diff
@@ -176,23 +176,82 @@ func VerifyEvidence(stateDB dbm.DB, state State, evidence types.Evidence) error
 			evidence.Time(), state.LastBlockTime.Add(evidenceParams.MaxAgeDuration))
 	}
 
+	if ev, ok := evidence.(*types.LunaticValidatorEvidence); ok {
+		switch ev.InvalidHeaderField {
+		case "ValidatorsHash":
+			if bytes.Equal(committedHeader.ValidatorsHash, ev.Header.ValidatorsHash) {
+				return errors.New("validators_hash matches committed hash")
+			}
+		case "NextValidatorsHash":
+			if bytes.Equal(committedHeader.NextValidatorsHash, ev.Header.NextValidatorsHash) {
+				return errors.New("next_validators_hash matches committed hash")
+			}
+		case "ConsensusHash":
+			if bytes.Equal(committedHeader.ConsensusHash, ev.Header.ConsensusHash) {
+				return errors.New("consensus_hash matches committed hash")
+			}
+		case "AppHash":
+			if bytes.Equal(committedHeader.AppHash, ev.Header.AppHash) {
+				return errors.New("app_hash matches committed hash")
+			}
+		case "LastResultsHash":
+			if bytes.Equal(committedHeader.LastResultsHash, ev.Header.LastResultsHash) {
+				return errors.New("last_results_hash matches committed hash")
+			}
+		default:
+			return errors.New("unknown invalid_header_field")
+		}
+	}
+
 	valset, err := LoadValidators(stateDB, evidence.Height())
 	if err != nil {
 		// TODO: if err is just that we cant find it cuz we pruned, ignore.
 		// TODO: if its actually bad evidence, punish peer
 		return err
 	}
 
-	// The address must have been an active validator at the height.
-	// NOTE: we will ignore evidence from H if the key was not a validator
-	// at H, even if it is a validator at some nearby H'
-	// XXX: this makes lite-client bisection as is unsafe
-	// See https://github.com/tendermint/tendermint/issues/3244
-	ev := evidence
-	height, addr := ev.Height(), ev.Address()
-	_, val := valset.GetByAddress(addr)
-	if val == nil {
-		return fmt.Errorf("address %X was not a validator at height %d", addr, height)
+	addr := evidence.Address()
+	var val *types.Validator
+
+	if _, ok := evidence.(*types.PhantomValidatorEvidence); ok {
+		_, val = valset.GetByAddress(addr)
+		if val != nil {
+			return fmt.Errorf("address %X is a validator at height %d", addr, evidence.Height())
+		}
+
+		// TODO: something smarter here
```

Comment:
Is there perhaps a list of validators that currently have something staked that you could loop through instead?

Reply:
TM does not have a notion of stake. But we can track validators within `evidence.MaxAge`